AX: Ensure native HTML radio buttons on macOS maintain accurate group count when group membership changes

rahimabdi opened the pull request:
#### 5b80ea62b394ef3c268ba1c2b7ef929d8d7ffb9c
<pre>
AX: Ensure native HTML radio buttons on macOS maintain accurate group count when group membership changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=297929">https://bugs.webkit.org/show_bug.cgi?id=297929</a>
<a href="https://rdar.apple.com/159221583">rdar://159221583</a>

Reviewed by Tyler Wilcock.

When the membership of a native HTML radio group changes (e.g., adding one or more radio buttons to a group via user interaction),
AccessibilityNodeObject::radioButtonGroup() determines group membership when the AX object is first created. However,
any membership changes must also be propagated to the platform accessibility API so that the backing AX object used for querying
current AX state (mirrored in the AX isolated tree) is kept up-to-date.

This PR ensures that upon radio group membership change, a new AXNotification::RadioGroupMembershipChanged is dispatched for every
radio in the affected group. In this manner, assistive technologies will convey the accurate group count &quot;X of Y&quot; for existing members,
where X is the radio&apos;s current position in the group and Y is the group count.

Note: the cardinality of radio buttons is not derived from AXProperty::PosInSet nor AXProperty::SetSize because these attributes are
not queried by the platform accessibility API for this purpose. For native radio buttons, the &quot;X of Y&quot; assistive technology
announcement is exposed via NSAccessibilityLinkedUIElementsAttribute which calls into
AXCoreObject::linkedObjects().

* LayoutTests/accessibility/radio-button-group-dynamic-changes-expected.txt: Added.
* LayoutTests/accessibility/radio-button-group-dynamic-changes.html: Added.
* LayoutTests/platform/glib/TestExpectations: skip radio-button-group-dynamic-changes.html.

* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onRadioGroupMembershipChanged):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/dom/RadioButtonGroups.cpp:
(WebCore::RadioButtonGroups::addButton):
(WebCore::RadioButtonGroups::removeButton):

Canonical link: <a href="https://commits.webkit.org/299585@main">https://commits.webkit.org/299585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca2475d4821132bf54988ee8cafe15e9c7279a2a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119437 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29779 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125706 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71508 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121314 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39821 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47705 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90755 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60060 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122389 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31781 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107087 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71232 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30817 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25196 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69345 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101237 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25387 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128682 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46355 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35088 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99330 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103286 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99141 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25200 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44576 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22581 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42924 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46218 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51918 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45683 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49033 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47370 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/radio-button-group-dynamic-changes.html:
```diff
@@ -0,0 +1,61 @@
+<script src="../resources/js-test.js"></script>
+<script>
+    if (window.accessibilityController)
+        window.jsTestIsAsync = true;
```
We typically put `window.jsTestIsAsync = true;` just above `var output`.

twilco commented on LayoutTests/accessibility/radio-button-group-dynamic-changes.html:
```diff
@@ -0,0 +1,61 @@
+<div id ="content">
+    <input id="radio1" type="radio" name="group1">
+    <input id="radio2" type="radio" name="group1">
+    <input id="radio2a" type="radio" name="group2" aria-label="first">
```
Is this `aria-label` necessary for the purpose of this test?

rahimabdi replied:
Ahh, I don't think so. I gave the radio buttons accessible names to better understand `debug(radio.attributesOfLinkedUIElements())` but unnecessary to keep so I've removed them.

twilco commented on LayoutTests/accessibility/radio-button-group-dynamic-changes.html:
```diff
@@ -0,0 +1,61 @@
+async function focusRadio2() {
+    document.getElementById("radio2").checked = false;
+    document.getElementById("radio2").click();
+}
```
Does this need to be an `async` function? I don't see any `await` usage inside, so I'd assume it could just be a normal function 🤔

rahimabdi replied:
Good catch, I don't believe so.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1939,6 +1939,21 @@ void AXObjectCache::onDetailsSummarySlotChange(const HTMLDetailsElement& details
 
+void AXObjectCache::radioGroupMembershipChanged(HTMLElement& radio)
+{
+    if (auto radioElement = dynamicDowncast<HTMLInputElement>(radio)) {
```
We try not to use `auto` for pointer types. `dynamicDowncast<T>` returns a `T*`, being nullptr if the cast failed because the instance was not a `T`. So the proper type would be `auto*`.

rahimabdi replied:
Got it, updated to `auto*`. 

Alternatively, what are your thoughts on using `RefPtr<HTMLInputElement>` here? (I assume slightly more overhead with refcounting but clearer?)

twilco replied:
I think we should use `RefPtr` (no template type specification), but if and only if the safer-cpp bot requires it. Otherwise `auto*` is best.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1939,6 +1939,21 @@ void AXObjectCache::onDetailsSummarySlotChange(const HTMLDetailsElement& details
+            if (auto* axObject = getOrCreate(sibling.ptr()))
+                postNotification(axObject, &sibling->document(), AXNotification::RadioGroupMembershipChanged);
+        }
+    }
```
Here we do an O(n) traversal to remove `radio` from the group, then another O(n) traversal to post the notifications. Instead, maybe we could do something like this, which is one O(n) traversal:

```cpp
for (auto& sibling : siblings) {
    if (sibling == &radio)
        continue;

    if (auto* axObject = getOrCreate(sibling.ptr()))
        postNotification(axObject, &sibling->document(), AXNotification::RadioGroupMembershipChanged);
}
```

Also, using `get` over `getOrCreate` is probably better here. We're posting a notification for the sole purpose of updating an accessibility object, but if one hasn't been created yet, there's nothing to update.

rahimabdi replied:
Right! I've updated but looks like I need to unwrap via `sibling.ptr()` to perform equality check with the `HTMLInputElement*` raw pointer since it's a `<Ref<HTMLInputElement>`. Thank you.

twilco commented on Source/WebCore/dom/RadioButtonGroups.cpp:
```diff
@@ -213,6 +214,9 @@ void RadioButtonGroups::addButton(HTMLInputElement& element)
     group->add(element);
+
+    if (CheckedPtr axCache = element.document().existingAXObjectCache())
+        axCache->radioGroupMembershipChanged(element);
```
We typically go for the shorter `cache` rather than `axCache` in little statements like these.

twilco commented on Source/WebCore/dom/RadioButtonGroups.cpp:
```diff
@@ -287,6 +291,9 @@ void RadioButtonGroups::removeButton(HTMLInputElement& element)
         m_nameToGroupMap.remove(it);
+
+    if (CheckedPtr axCache = element.document().existingAXObjectCache())
+        axCache->radioGroupMembershipChanged(element);
```
Same comment about `axCache`.

twilco commented on LayoutTests/platform/glib/accessibility/radio-button-group-dynamic-changes-expected.txt:
```diff
@@ -0,0 +1,7 @@
+FAIL successfullyParsed should be true. Was false.
+Some tests failed.
+
+TEST COMPLETE
```
It's typically best not to check in failures for non-WPTs. Linked UI elements is mostly a macOS / iOS concept I believe, so we should just skip this test for glib platforms.

rahimabdi replied:
Makes sense, thanks.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1939,6 +1939,20 @@ void AXObjectCache::onDetailsSummarySlotChange(const HTMLDetailsElement& details
+    if (auto* radioElement = dynamicDowncast<HTMLInputElement>(radio)) {
+        auto siblings = radioElement->radioButtonGroup();
+
+        for (auto& sibling : siblings) {
```
We can shorten this to:

```cpp
for (auto& sibling : radioElement->radioButtonGroup()) {
```

twilco commented on Source/WebCore/dom/RadioButtonGroups.cpp:
```diff
@@ -213,6 +214,9 @@ void RadioButtonGroups::addButton(HTMLInputElement& element)
     group->add(element);
+
+    if (CheckedPtr cache = element.document().existingAXObjectCache())
+        cache->radioGroupMembershipChanged(element);
```
Sorry, forgot to mention this in the last round, but for methods that exist to notify us of DOM changes, we use an "on" prefix. So here, this would be `onRadioGroupMembershipChanged`

rahimabdi replied:
Copy!

rahimabdi committed:
AX: Ensure native HTML radio buttons on macOS maintain accurate group count when group membership changes
https://bugs.webkit.org/show_bug.cgi?id=297929
rdar://159221583

Reviewed by Tyler Wilcock.

When the membership of a native HTML radio group changes (e.g., adding one or more radio buttons to a group via user interaction),
AccessibilityNodeObject::radioButtonGroup() determines group membership when the AX object is first created. However,
any membership changes must also be propagated to the platform accessibility API so that the backing AX object used for querying
current AX state (mirrored in the AX isolated tree) is kept up-to-date.

This PR ensures that upon radio group membership change, a new AXNotification::RadioGroupMembershipChanged is dispatched for every
radio in the affected group. In this manner, assistive technologies will convey the accurate group count "X of Y" for existing members,
where X is the radio's current position in the group and Y is the group count.

Note: the cardinality of radio buttons is not derived from AXProperty::PosInSet nor AXProperty::SetSize because these attributes are
not queried by the platform accessibility API for this purpose. For native radio buttons, the "X of Y" assistive technology
announcement is exposed via NSAccessibilityLinkedUIElementsAttribute which calls into
AXCoreObject::linkedObjects().

* LayoutTests/accessibility/radio-button-group-dynamic-changes-expected.txt: Added.
* LayoutTests/accessibility/radio-button-group-dynamic-changes.html: Added.
* LayoutTests/platform/glib/TestExpectations: skip radio-button-group-dynamic-changes.html.

* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onRadioGroupMembershipChanged):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/dom/RadioButtonGroups.cpp:
(WebCore::RadioButtonGroups::addButton):
(WebCore::RadioButtonGroups::removeButton):

Canonical link: https://commits.webkit.org/299585@main
